### DISCUSSION TITLE
How to deal with expected errors?

### DISCUSSION THREAD

**Discussion by ThomasTr** [symfony/symfony#57465] [Q&A]
I have one or more places in my code that can fail as expected, e.g. the call to file_get_contents(); I intercept these with a try / catch. I use safe/file_get_contents here:

```

use function Safe\file_get_contents;

try
{
    $contents = file_get_contents($match);
}
catch (\ErrorException)
{
    continue;
}
```

However, file_get_contents emits E_Warnings, which I cannot intercept / suppress in this way. However, [this change](https://github.com/symfony/symfony/commit/7e4f02e13e59c5e813f3c3fa1eab865f490f9b62) means that all warnings are now treated as errors and trigger corresponding emails, as I want to be notified of serious errors.

Unfortunately, this also removes the possibility of deliberately suppressing errors. For now I will overwrite the above change in my config.

In your opinion, is there a better way to deal with expected errors without them appearing in the logs and triggering corresponding routines? I don't really want to suppress errors with @.

**Comment by Trismegiste** [comment]
I agree : suppressing errors is a terrible idea.

Why your ```file_get_contents``` emits warning in the first place ?

**Comment by ThomasTr** [comment]
In this case, an attempt is made to convert an external image to an inline image. This sometimes fails if, for example, the url is temporarily unavailable or is generally invalid. However, this is not a problem and can be ignored.

**Reply by Trismegiste** [reply]
Loading a picture with ```file_get_contents``` is a terrible idea because the picture will be fully loaded in memory.
Those kind of contents should be streamed with fpassthru, at the minimum

If the URL could fail, you have to build a failover (at least a generic broken picture) or your users would be puzzled. 

For your problem, you can check if the URL is ok with ```get_headers()``` php function

**Comment by stof** [comment]
The commit you linked is only changing the logging level. It does not change the error handler behavior at all.

And the Symfony error handler respects the `@` operator to silence warnings.

Side note: Symfony throws an ErrorException for warnings only in debug mode. So catching those ErrorException is almost always a mistake (as that would not work in prod)

**Comment by ThomasTr** [comment]
>Side note: Symfony throws an ErrorException for warnings only in debug mode. So catching those ErrorException is almost always a mistake (as that would not work in prod)

I'm using here Safe\file_get_contents which throws FilesystemException which extends \ErrorException. Originally, I had also caught the FilesystemException here once. Switching to \ErrorException was part of my troubleshooting.

The logging level is exactly my problem: However, file_get_contents also emits warnings. By changing the logging level, these become errors and thus, in my case, part of my error notification routine.

With the above LogLevel change in Symfony 7.1 that turns warnings into errors, there is no way to intentionally catch errors for php functions that also emit warnings. The warnings then appear in the logs as errors.

**Reply by stof** [reply]
> The logging level is exactly my problem: However, file_get_contents also emits warnings. By changing the logging level, these become errors and thus, in my case, part of my error notification routine.

This is the part that I don't understand. logging does not convert warnings to errors at all. And logging is not about catching errors or warnings (at least not in the same meaning of catching than your code snippet using `try/catch`).

And if you are talking only about the error notifications of your log monitoring system, you are still free to change the configuration by configuring different log levels: https://symfony.com/doc/current/reference/configuration/framework.html#php-errors
What this commit changed is only the _default_ configuration.

**Comment by ThomasTr** [comment]
Many thanks for your suggestions.
